src/catalog/src/durable/upgrade.rs

```diff
@@ -297,4 +313,74 @@ pub(crate) mod persist {
 
         Ok(upper)
     }
+
+    /// Runs `migration_logic` on the contents of the current catalog assuming a current version of
+    /// `current_version`.
+    ///
+    /// Returns the new version and upper.
+    async fn run_versioned_upgrade<V1: IntoStateUpdateKindBinary, V2: IntoStateUpdateKindBinary>(
+        persist_handle: &mut PersistHandle,
+        upper: Timestamp,
+        current_version: u64,
+        migration_logic: impl FnOnce(Vec<V1>) -> Vec<MigrationAction<V1, V2>>,
+    ) -> Result<(u64, Timestamp), CatalogError> {
+        // 1. Get a snapshot of the current catalog, using the current version to deserialize the
+        // contents.
```

Comment:
nit: "current" is a bit ambiguous here, it should be the old/`V1`/previous version, right?

Reply:
Good point, updated the comment.